src/python/dxpy/sugar/processing.py

```diff
@@ -0,0 +1,519 @@
+                "and file handles have been closed."
+            )
+        stderr = []
+        if self.capture_stderr:
+            stderr.extend(self._stderr_buffers)
+        if self._stderr_type in (_BUFFER, _PIPE):
+            stderr.append(self.err)
+        return stderr
+
+    @property
+    def was_run(self):
+        """
+        Whether the `run()` method was called.
+        """
+        return bool(self._processes)
+
+    @property
+    def done(self):
+        """
+        Whether the commands have finished running.
+        """
+        return self.returncode is not None
+
+    @property
+    def ok(self):
+        """
+        Whether the commands completed successfully, i.e. all had `returncode == 0`.
+        """
+        return self.returncode == 0
+
+    @property
+    def closed(self):
+        """
+        Whether `self.close()` has been called.
+        """
+        return self._closed
+
+    def run(self, echo=None, **kwargs):
+        """
+        Run the commands.
+
+        Args:
+            echo: Whether to echo the commands to the log. Overrides the value of
+                `self.echo`.
+            kwargs: Keyword arguments to pass to Popen. Overrides any values specified
+                in `self.popen_kwargs`.
+        """
+        if self.was_run:
+            raise RuntimeError("Cannot call run() more than once.")
+
+        if echo is None:
+            echo = self.echo
+        if echo is not False:
+            LOG.info(str(self))
+
+        num_commands = len(self.cmds)
+        procs = []
+
+        for i, cmd in enumerate(self.cmds, 1):
+            popen_kwargs = copy.copy(self.popen_kwargs)
+            popen_kwargs.update(kwargs)
+            if procs:
+                popen_kwargs["stdin"] = procs[-1].stdout
+            if i == num_commands:
+                popen_kwargs["stdout"] = self._init_stdout()
+                popen_kwargs["stderr"] = self._init_stderr()
+            else:
+                popen_kwargs["stdout"] = subprocess.PIPE
+                popen_kwargs["stderr"] = self._get_stderr_buffer()
+            proc = subprocess.Popen(cmd, **popen_kwargs)
+            if procs:
+                procs[-1].stdout.close()
+            procs.append(proc)
+
+        self._processes = procs
+
+    def block(self, close=True, raise_on_error=True):
+        """
+        Wait for all commands to finish.
+
+        Args:
+            close: Whether to call `self.close()` after the processes complete.
+            raise_on_error: Whether to raise a :class:`subprocess.CalledProcessError`
+                if the returncode was not 0.
+        """
+        last_proc = self._processes[-1]
+        try:
+            out, err = (
+                "" if std is None else std.strip()
+                for std in last_proc.communicate()
+            )
+            if self._stdout_type == _PIPE:
+                self.out = out
+            if self._stderr_type == _PIPE:
+                self.err = err
```

Comment:
I don't like the use of .out and .err here. This is mostly aesthetic but the command call "run_cmd(cmd).out" seems pretty unintuitive to me. Perhaps we can add some helper functions, such as "get_stdout()" and "get_stderr()" and have those return .out and .err?

Reply:
This would make the command: `run_cmd(cmd).get_stdout()`

Reply:
I cleaned this up a bit, and changed to providing these values via the 'output' and 'error' properties rather than allowing direct access to the attributes.